extensions/amp-analytics/0.1/variables.js

```diff
@@ -277,24 +266,68 @@ export class VariableService {
       let value = options.getVar(name);
 
       if (typeof value == 'string') {
-        value = this.expandTemplateSync(
+        value = this.expandTemplate(
           value,
           new ExpansionOptions(
             options.vars,
             options.iterations - 1,
             true /* noEncode */
-          )
+          ),
+          element
         );
       }
 
-      if (!options.noEncode) {
-        value = encodeVars(/** @type {string|?Array<string>} */ (value));
-      }
-      if (value) {
-        value += argList;
-      }
-      return value;
+      const bindings = this.getMacros(element);
+      const urlReplacements = Services.urlReplacementsForDoc(element);
+      const whitelist = element.hasAttribute('sandbox')
+        ? SANDBOX_AVAILABLE_VARS
+        : undefined;
+
+      return Promise.resolve(value)
+        .then(value => {
+          if (isArray(value)) {
+            return Promise.all(
+              value.map(item =>
+                urlReplacements.expandStringAsync(item, bindings, whitelist)
+              )
+            );
+          }
+          return urlReplacements.expandStringAsync(
+            value + argList,
+            bindings,
+            whitelist
+          );
+        })
+        .then(value => {
+          if (!options.noEncode) {
+            value = encodeVars(/** @type {string|?Array<string>} */ (value));
+          }
+          return value;
+        });
+    });
+  }
+
+  /**
+   * Wrapper around String.replace that handles asynchronous resolution.
+   * @param {string} str
+   * @param {RegExp} regex
+   * @param {Function} replacer
+   * @return {!Promise<string>}
+   */
+  asyncStringReplace_(str, regex, replacer) {
```

Comment:
If I understand the code correctly. What we want here to an array of promise that waits for each variable to resolve. Personally I think this helper method makes the code more difficult to understand. 
Can we inline the code as previous and do 
const promises = []
str.replace(regex, () => {
  // some operation
  promises.push(promise)
})
return Promise.all(promises)
```

So that we can get rid of the stringBuilder here. Thanks